store/v2/root/snapshot.go

```diff
@@ -0,0 +1,213 @@
+package root
+
+import (
+	"bufio"
+	"compress/zlib"
+	"fmt"
+	prefixdb "github.com/cosmos/cosmos-sdk/db/prefix"
+	"github.com/cosmos/cosmos-sdk/snapshots"
+	snapshottypes "github.com/cosmos/cosmos-sdk/snapshots/types"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	types "github.com/cosmos/cosmos-sdk/store/v2"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	protoio "github.com/gogo/protobuf/io"
+	"io"
+	"math"
+)
+
+// Restore implements snapshottypes.Snapshotter.
+func (rs *Store) Restore(height uint64, format uint32, chunks <-chan io.ReadCloser, ready chan<- struct{}) error {
+	if height == 0 {
+		return sdkerrors.Wrap(sdkerrors.ErrLogic, "cannot restore snapshot at height 0")
+	}
+	if height > uint64(math.MaxInt64) {
+		return sdkerrors.Wrapf(snapshottypes.ErrInvalidMetadata,
+			"snapshot height %v cannot exceed %v", height, int64(math.MaxInt64))
+	}
+
+	versions, err := rs.stateDB.Versions()
+	if versions.Count() != 0 {
+		return sdkerrors.Wrapf(sdkerrors.ErrLogic, "cannot restore snapshot for non empty store at height %v", height)
+	}
+
+	// Signal readiness. Must be done before the readers below are set up, since the zlib
+	// reader reads from the stream on initialization, potentially causing deadlocks.
+	if ready != nil {
+		close(ready)
+	}
+
+	// Set up a restore stream pipeline
+	// chan io.ReadCloser -> chunkReader -> zlib -> delimited Protobuf -> ExportNode
+	chunkReader := snapshots.NewChunkReader(chunks)
+	defer chunkReader.Close()
+	zReader, err := zlib.NewReader(chunkReader)
+	if err != nil {
+		return sdkerrors.Wrap(err, "zlib failure")
+	}
+	defer zReader.Close()
+	protoReader := protoio.NewDelimitedReader(zReader, snapshotMaxItemSize)
+	defer protoReader.Close()
+
+	var subStore *substore
+
+	for {
+		item := &storetypes.SnapshotItem{}
+		err := protoReader.ReadMsg(item)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return sdkerrors.Wrap(err, "invalid protobuf message")
+		}
+
+		switch item := item.Item.(type) {
+		case *storetypes.SnapshotItem_Schema:
+			schemaWriter := prefixdb.NewPrefixWriter(rs.stateTxn, schemaPrefix)
+			sKeys := item.Schema.GetKeys()
+			for _, sKey := range sKeys {
+				rs.schema[string(sKey)] = types.StoreTypePersistent
+				err := schemaWriter.Set(sKey, []byte{byte(types.StoreTypePersistent)})
+				if err != nil {
+					return sdkerrors.Wrap(err, "error at set the store schema key values")
+				}
+			}
+
+		case *storetypes.SnapshotItem_Store:
+			storeName := item.Store.GetName()
+			// checking the store schema exists or not
+			if _, has := rs.schema[storeName]; !has {
+				return sdkerrors.Wrapf(sdkerrors.ErrLogic, "received store name before store schema %s", storeName)
```

Comment:
still should update this so we have separate checks for 1) schema not received yet and 2) store is missing from schema

Reply:
I Can use storeSchemaReceived  bool to check the schema is received or not 

```go
	

var storeSchemaReceived bool = false


switch item := item.Item.(type) {
case *storetypes.SnapshotItem_Schema:
   storeSchemaReceived = true
			
case *storetypes.SnapshotItem_Store:
  storeName := item.Store.GetName()
  // checking the store schema is received or not
  if !storeSchemaReceived {
      return sdkerrors.Wrapf(sdkerrors.ErrLogic, "received store name before store schema %s", storeName)
  }
  // checking the store schema exists or not
  if _, has := rs.schema[storeName]; !has {
       return sdkerrors.Wrapf(sdkerrors.ErrLogic, "store is missing from schema %s", storeName)
  }
}
```